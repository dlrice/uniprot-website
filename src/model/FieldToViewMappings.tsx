import React from 'react';
import SimpleView from './SimpleView';
import { ProteinNames, ProteinNamesData } from './ProteinNames';
import { GeneNames, GeneNamesData } from './GeneNames';
import { Organism, OrganismData } from './Organism';

const FieldToViewMappings: {
  [index: string]: (row: any) => JSX.Element | undefined;
} = {
  accession: (row: { primaryAccession: string }) => (
    <SimpleView
      termValue={row.primaryAccession}
      linkTo={`/uniprotkb/${row.primaryAccession}`}
    />
  ),
  id: (row: { uniProtId: string }) => <SimpleView termValue={row.uniProtId} />,
  protein_name: (data: ProteinNamesData) => {
    return <ProteinNames data={data} />;
  },
  gene_names: (data: GeneNamesData) => {
    return <GeneNames data={data} />;
  },
  organism: (data: OrganismData) => {
    return <Organism data={data} />;
  },
};

export default FieldToViewMappings;
