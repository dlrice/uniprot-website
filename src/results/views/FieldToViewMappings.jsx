import React from 'react';
import get from 'lodash.get';
import SimpleView from './SimpleView';
import NameView from './NameView';

const FieldToViewMappings = {
  accession: row => <SimpleView termValue={row.accession} />,
  id: row => <SimpleView termValue={row.id} />,
  protein_name: (row) => {
    const alternativeNames = [];
    const ecNumber = get(row, 'protein.recommendedName.ecNumber.value');
    if (ecNumber) {
      alternativeNames.push(ecNumber);
    }
    const alternativeName = get(row, 'protein.recommendedName.alternativeName');
    if (alternativeName) {
      alternativeNames.push(alternativeName);
    }
    const props = {
      name: get(row, 'protein.recommendedName.fullName.value'),
      shortName: get(row, 'protein.recommendedName.shortName.value'),
      alternativeNames,
    };
    return <NameView {...props} />;
  },
  gene_name: (row) => {
    const props = {
      name: row.gene ? row.gene[0].name.value : '',
      alternativeNames: [
        // ...row.gene[0].synonyms.map(syn => syn.value),
        // ...row.gene[0].orfNames.map(orf => orf.value),
      ],
    };
    return <NameView {...props} />;
  },
};

export default FieldToViewMappings;
