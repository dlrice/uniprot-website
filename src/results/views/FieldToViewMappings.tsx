import React from 'react';
import idx from 'idx';
import SimpleView from './SimpleView';
import NameView from './NameView';

type ProteinNameRow = {
  protein: {
    recommendedName: {
      ecNumber: {
        value: string;
      };
      fullName: {
        value: string;
      };
      shortName: {
        value: string;
      };
    };
    alternativeName: [{ fullName: { value: string } }];
  };
};

type GeneNameRow = {
  gene: [
    {
      name: {
        value: string;
      };
      synonyms: [{ value: string }];
      orfNames: [{ value: string }];
    }
  ];
};

type OrganismRow = {
  organism: {
    names: [
      {
        type: string;
        value: string;
      }
    ];
  };
};

const organismNameReducer = (type: string) => (
  acc: string,
  organismName: { type: string; value: string },
) => {
  if (organismName.type === type) {
    if (type === 'scientific') {
      return `${acc} ${organismName.value}`;
    }
    return `${acc} (${organismName.value})`;
  }
  return acc;
};

const FieldToViewMappings = {
  accession: (row: { accession: string }) => <SimpleView termValue={row.accession} />,
  id: (row: { id: string }) => <SimpleView termValue={row.id} />,
  protein_name: (row: ProteinNameRow) => {
    const alternativeNames: string[] = [];
    const ecNumber = idx(row, _ => _.protein.recommendedName.ecNumber.value);
    if (ecNumber) {
      alternativeNames.push(ecNumber);
    }
    const alternativeNameArray = idx(row, _ => _.protein.alternativeName);
    if (alternativeNameArray) {
      const alternativeName = alternativeNameArray.map(name => name.fullName.value);
      if (alternativeName.length) {
        alternativeNames.push(...alternativeName);
      }
    }
    const props = {
      name: idx(row, _ => _.protein.recommendedName.fullName.value),
      shortName: idx(row, _ => _.protein.recommendedName.shortName.value),
      alternativeNames,
    };
    return <NameView {...props} />;
  },
  gene_names: (row: GeneNameRow) => {
    const genes = idx(row, _ => _.gene);
    if (!genes || genes.length <= 0) {
      return;
    }
    // const name = genes.map(gene => idx(gene, _ => _.name.value)).join(', ');
    const names: string[] = [];
    const alternativeNames: string[] = [];
    genes.forEach((gene) => {
      if (gene.name) {
        names.push(gene.name.value);
      }
      if (gene.synonyms) {
        alternativeNames.push(...gene.synonyms.map(syn => syn.value));
      }
      if (gene.orfNames) {
        alternativeNames.push(...gene.orfNames.map(orf => orf.value));
      }
    });
    const nameString = names.join(', ');
    const props = { name: nameString, alternativeNames };
    return <NameView {...props} />;
  },
  organism: (row: OrganismRow) => {
    const names = idx(row, _ => _.organism.names);
    if (!names) {
      return;
    }
    const termValue = ['scientific', 'common', 'synonym'].reduce(
      (acc, organismName) => `${acc} ${names.reduce(organismNameReducer(organismName), '')}`,
      '',
    );
    return <SimpleView termValue={termValue} />;
  },
};

export default FieldToViewMappings;
