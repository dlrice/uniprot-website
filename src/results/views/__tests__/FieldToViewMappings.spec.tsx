import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import FieldToViewMappings from '../FieldToViewMappings';

configure({ adapter: new Adapter() });

describe('FieldToViewMappings', () => {
  test('should render protein_name', () => {
    const data = {
      proteinDescription: {
        recommendedName: {
          fullName: {
            value: 'My protein',
          },
        },
        shortNames: [
          {
            value: 'My prot',
          },
        ],
        ecNumbers: [
          {
            value: 'ec1234',
          },
        ],
        alternativeNames: [
          {
            fullName: {
              value: 'Some protein',
            },
          },
        ],
      },
    };
    const component = FieldToViewMappings.protein_name(data);
    const wrapper = shallow(component);
    expect(wrapper.debug()).toMatchSnapshot();
  });

  test('should render gene_names', () => {
    const data = {
      genes: [
        {
          geneName: {
            value: 'My gene',
          },
          synonyms: [{ value: 'synonym 1' }, { value: 'synonym 2' }],
          orfNames: [
            { value: 'orfNames 1' },
            { value: 'orfNames 2' },
            { value: 'orfNames 3' },
          ],
        },
      ],
    };
    const component = FieldToViewMappings.gene_names(data);
    const wrapper = shallow(component);
    expect(wrapper.debug()).toMatchSnapshot();
  });
});
