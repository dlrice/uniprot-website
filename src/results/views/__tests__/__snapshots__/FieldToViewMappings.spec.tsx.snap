// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FieldToViewMappings should render gene_names 1`] = `
"<div>
  <strong>
    My gene
  </strong>
   
   
  <span>
    <span>
       
      synonym 1
    </span>
    <span>
       
      synonym 2
    </span>
    <span>
       
      orfNames 1
    </span>
    <span>
       
      orfNames 2
    </span>
    <span>
       
      orfNames 3
    </span>
  </span>
</div>"
`;

exports[`FieldToViewMappings should render protein_name 1`] = `
"<div>
  <strong>
    My protein
  </strong>
   
  My prot
   
  <span>
    <span>
       
      ec1234
    </span>
  </span>
</div>"
`;
